velox/functions/prestosql/ArrayDuplicates.cpp

```diff
@@ -0,0 +1,242 @@
+/// internal elements() vector from the original ArrayVector.
+///
+/// First a new vector is created containing the indices of the elements
+/// which will be present in the output, and wrapped into a DictionaryVector.
+/// Next the `lengths` and `offsets` vectors that control where output arrays
+/// start and end are wrapped into the output ArrayVector.
+template <typename T>
+class ArrayDuplicatesFunction : public exec::VectorFunction {
+ public:
+  // Execute function.
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      exec::Expr* caller,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args.front()->as<ArrayVector>();
+    auto elementsVector = arrayVector->elements();
+    ArrayDuplicatesFunction::validateElementType(elementsVector->typeKind());
+
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+    vector_size_t elementsCount = elementsRows.size();
+    vector_size_t rowCount = arrayVector->size();
+
+    // Allocate new vectors for indices, length and offsets.
+    memory::MemoryPool* pool = context->pool();
+    BufferPtr newIndices =
+        AlignedBuffer::allocate<vector_size_t>(elementsCount, pool);
+    BufferPtr newLengths =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+    BufferPtr newOffsets =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+
+    // Pointers and cursors to the raw data.
+    vector_size_t indicesCursor = 0;
+    auto* rawNewIndices = newIndices->asMutable<vector_size_t>();
+    auto* rawSizes = newLengths->asMutable<vector_size_t>();
+    auto* rawOffsets = newOffsets->asMutable<vector_size_t>();
+
+    // Convert coercible-numeric-typed ArrayVector to bigint ArrayVector
+    int64_t* bigintResultRawValues;
+    std::shared_ptr<BaseVector> bigintResultElementVector = nullptr;
+    constexpr bool isCoercibleType = std::is_same_v<T, int8_t> ||
+        std::is_same_v<T, int16_t> || std::is_same_v<T, int32_t>;
+    if constexpr (isCoercibleType) {
+      auto bigintResultArray = (*result)->as<ArrayVector>();
+      bigintResultElementVector = bigintResultArray->elements();
+      bigintResultElementVector->resize(elementsCount);
+      auto resultFlatElements =
+          bigintResultElementVector->asFlatVector<int64_t>();
+      bigintResultRawValues = resultFlatElements->mutableRawValues();
+    }
+
+    // Process the rows: use a hashmap to store unique values and
+    // whether it occurred once or more than once.
+    folly::F14FastMap<T, bool> uniqueMap;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto size = arrayVector->sizeAt(row);
+      auto offset = arrayVector->offsetAt(row);
+
+      rawOffsets[row] = indicesCursor;
+      vector_size_t numOfNulls = 0;
```

Comment:
inconsistent naming: numOfNulls vs. rowCount and elementsCount. Elsewhere in the codebase we use numXxx, e.g. numNulls, numRows and numElements.